File: pacs_photproject.py
#!/usr/bin/env python2.6
#
# NAME: pacs_photproject
# 
# DESCRIPTION: create a map from a set of PACS observations, by backprojecting 
# the timelines onto the sky map, and by dividing it by the weights, which are
# the backprojection of 1.
# The routine is meant to replicate HCSS' photproject
#
# Author: P. Chanial

from   matplotlib.pyplot import figure, plot, show
import numpy
from   optparse import OptionParser
import sys
import tamasisfortran as tmf
from   tamasis import Map, PacsObservation, PacsProjectionSharpEdges

# set up the option parser
parser = OptionParser('Usage: %prog [options] fitsfile...')
parser.add_option('-o', help='write output map to FILE in FITS format [default:'
                  ' %default]', metavar='FILE', dest='outputfile',     default=
                  'photproject.fits')
parser.add_option('--header', help='use FITS header in FILE to specify the map '
                  'projection [default: automatic]', metavar='FILE')
parser.add_option('--resolution', help='input pixel size of the map in arcsecon'
                  'ds [default: %default]', default=3.2)
parser.add_option('-n', '--npixels-per-sample', help='Maximum number of sky pix'
                  'els intercepted by a PACS detector [default: %default]', 
                  default=6)
parser.add_option('--no-flatfield', help='do not divide by calibration flat-fie'
                  'ld [default: False]', dest='do_flatfielding', action='store_'
                  'false', default=True)
parser.add_option('-f', '--filtering', help='method for timeline filtering: mea'
                  'n or none [default: %default]', metavar='METHOD', default='m'
                  'ean')
parser.add_option('-d', '--deglitching', help='method for timeline deglitching:'
                  ' l2std, l2mad or none [default: %default]', metavar='METHOD',
                  default='none')
parser.add_option('--nsigma-deglitching', help='N-sigma deglitching value [defa'
                  'ult: %default]', dest='nsigma', default=5.)
parser.add_option('--plot', help='plot the map', action='store_true', dest='do_'
                  'plot', default=False)

(options, filename) = parser.parse_args(sys.argv[1:])

if len(filename) == 0:
    raise SystemExit(parser.print_help() or 1)

# Check options
options.deglitching = options.deglitching.lower()
if options.deglitching not in ('none', 'l2std', 'l2mad'):
    raise ValueError("Invalid deglitching method '"+options.deglitching+"'. Val"
                     "id methods are 'l2std', 'l2mad' or 'none'.")

options.filtering = options.filtering.lower()
if options.filtering not in ('none', 'mean'):
    raise ValueError("Invalid filtering method '"+options.filtering+"'. Valid m"
                     "ethods are 'mean', or 'none'.")

# Set up the PACS observation(s)
pacs = PacsObservation(filename=filename,
                       header=options.header,
                       resolution=options.resolution,
                       fine_sampling_factor=1,
                       keep_bad_detectors=False,
                       npixels_per_sample=options.npixels_per_sample)

# Get map dimensions
nx = pacs.header['naxis1']
ny = pacs.header['naxis2']

# Set up the acquisition model. finer_sampling is set to False because
# photproject does not attempt to sample better than what is transmitted
projection = PacsProjectionSharpEdges(pacs, finer_sampling=False)

# Read the timeline
tod = pacs.get_tod(do_flatfielding=options.do_flatfielding, do_subtraction_mean=False)

if options.filtering == 'mean':
    print 'Removing mean value...'
    tod -= tod.mean(axis=0)

# Deglitch
if options.deglitching != 'none':
    nbads = numpy.sum(tod.mask)
    if options.deglitching == 'l2std':
        tod.mask = tmf.deglitch_l2b_std(projection.pmatrix, nx, ny, tod, 
            tod.mask.astype('int8'), options.nsigma, pacs.npixels_per_sample)
    else:
        tod.mask = tmf.deglitch_l2b_mad(projection.pmatrix, nx, ny, tod, 
            tod.mask.astype('int8'), options.nsigma, pacs.npixels_per_sample)
    print 'Number of glitches detected:', numpy.sum(tod.mask) - nbads

# Backproject the timeline and divide it by the weight
print 'Computing the map...'
mymap = Map.zeros((nx, ny), order='f', header=pacs.header)
tmf.backprojection_weighted(projection.pmatrix, tod, tod.mask.astype('int8'), 
                            mymap, pacs.npixels_per_sample)

# Write resulting map as a FITS file
mymap.writefits(options.outputfile)

# Plot the map
if options.do_plot:
     mymap.imshow()
     show()
#    idetector = 0
#    figure()
#    plot(tod[:,idetector])
#    index=numpy.where(tod.mask[:,idetector])
#    plot(index, tod.data[index,idetector], 'ro')
#    show()

File: pacs_photproject.f
program pacs_photproject

    use iso_fortran_env,        only : ERROR_UNIT, OUTPUT_UNIT
    use module_fitstools,       only : ft_header2str, ft_read_parameter, ft_write
    use module_deglitching,     only : deglitch_l2b
    use module_optionparser,    only : optionparser
    use module_pacsinstrument,  only : pacsinstrument
    use module_pacsobservation, only : pacsobservation, maskarray
    use module_pointingmatrix,  only : backprojection_weighted, pointingelement
    use module_preprocessor,    only : divide_vectordim2, median_filtering_nocopy, subtract_meandim1
    use module_string,          only : strlowcase
    use module_tamasis,         only : init_tamasis
    implicit none

    class(pacsinstrument), allocatable :: pacs
    class(pacsobservation), allocatable:: obs
    type(maskarray)                    :: maskarray_policy
    type(pointingelement), allocatable :: pmatrix(:,:,:)
    class(optionparser), allocatable   :: parser
    real*8, allocatable                :: signal(:,:)
    logical*1, allocatable             :: mask(:,:)
    real*8, allocatable                :: map1d(:)
    character(len=2048), allocatable   :: infile(:)
    character(len=2048)                :: outfile, headerfile
    character(len=2880)                :: header
    real*8                             :: resolution
    integer                            :: npixels_per_sample
    logical                            :: do_flatfield
    character(len=256)                 :: filtering_method, deglitching_method
    integer*8                          :: nsamples, idetector
    integer                            :: nobs, iobs, nx, ny, start
    integer                            :: status, count1, count2, count_rate, count_max
    integer                            :: filtering_length
    real*8                             :: deglitching_nsigma

    ! command parsing
    allocate(parser)
    call parser%init('pacs_photproject [options] fitsfile...', 1, -1)
    call parser%add_option('', 'o', 'Filename of the output map (FITS format)', has_value=.true., default='photproject.fits')
    call parser%add_option('header', '', 'Input FITS header of the map',        has_value=.true.)
    call parser%add_option('resolution', '', 'Input pixel size of the map',     has_value=.true., default='3.2')
    call parser%add_option('npixels-per-sample', 'n', 'Maximum number of sky pixels intercepted by a PACS detector',               &
                           has_value=.true.)
    call parser%add_option('no-flatfield','','Do not divide by calibration flat field')
    call parser%add_option('filtering', 'f', 'Timeline filtering (median|none)',  has_value=.true., default='none')
    call parser%add_option('length', 'l', 'Filtering length',  has_value=.true., default='200')
    call parser%add_option('deglitching', 'd', 'Timeline deglitching (l2std|l2mad|none)', has_value=.true., default='none')
    call parser%add_option('nsigma', '', 'N-sigma for deglitching',             has_value=.true., default='5.')

    call parser%parse(status)
    if (status == -1) stop
    if (status /=  0) go to 999

    outfile = parser%get_option('o') !XXX status gfortran bug
    headerfile = parser%get_option('header')
    resolution = parser%get_option_as_real('resolution')
    npixels_per_sample = parser%get_option_as_integer('npixels-per-sample')
    do_flatfield       = .not. parser%get_option_as_logical('no-flatfield')
    filtering_method   = strlowcase(parser%get_option('filtering'))
    filtering_length   = parser%get_option_as_integer('length')
    deglitching_method = strlowcase(parser%get_option('deglitching'))
    deglitching_nsigma = parser%get_option_as_real('nsigma')

    if (filtering_method /= 'none' .and. filtering_method /= 'median') then
        write (ERROR_UNIT) "Invalid filtering method '" // filtering_method // "'. The only valid method is 'median'."
        go to 999
    end if
    
    if (deglitching_method /= 'none' .and. deglitching_method /= 'l2std' .and. deglitching_method /= 'l2mad') then
        write (ERROR_UNIT) "Invalid deglitching method '" // deglitching_method // "'. Valid methods are 'l2std' or 'l2mad'."
        go to 999
    end if
    
    write (OUTPUT_UNIT,*)

    ! initialise tamasis
    call init_tamasis

    ! initialise observations
    nobs = parser%get_argument_count()
    allocate(obs)
    call parser%get_arguments(infile, status)
    if (status /= 0) go to 999
    call obs%init(infile, maskarray_policy, status, verbose=.true.)
    if (status /= 0) go to 999
    nsamples = obs%nsamples_tot

    ! print observation info
    call obs%print()

    ! get npixels_per_sample
    if (parser%get_option('npixels-per-sample') == '') then
        if (obs%channel == 'r') then
            npixels_per_sample = 11
        else
            npixels_per_sample = 5
        endif
    else
        npixels_per_sample = parser%get_option_as_integer('npixels-per-sample')
    end if

    ! initialise pacs instrument
    allocate(pacs)
    call pacs%init(obs%channel, obs%observing_mode == 'Transparent', 1, .false., status)
    if (status /= 0) go to 999

    ! get FITS header
    if (headerfile /= '') then
       call ft_header2str(headerfile, header, status)
    else
       call pacs%compute_map_header(obs, .false., resolution, header, status)
    end if
    if (status /= 0) go to 999

    ! allocate memory for the maps
    call ft_read_parameter(header, 'naxis1', nx, count2, status=status)
    if (status /= 0 .or. count2 == 0) go to 999

    call ft_read_parameter(header, 'naxis2', ny, count2, status=status)
    if (status /= 0 .or. count2 == 0) go to 999

    allocate(map1d(0:nx*ny-1))

    ! compute the projector
    allocate (pmatrix(npixels_per_sample,nsamples,pacs%ndetectors))
    call pacs%compute_projection_sharp_edges(obs, .false., header,  nx, ny, pmatrix, status)
    if (status /= 0) go to 999

    ! read the signal file
    allocate (signal(nsamples, pacs%ndetectors))
    allocate (mask  (nsamples, pacs%ndetectors))
    call pacs%read(obs, signal, mask, status, verbose=.true.)
    if (status /= 0) go to 999

    ! remove flat field
    if (do_flatfield) then
        write (OUTPUT_UNIT,'(a)') 'Flat-fielding... '
        call divide_vectordim2(signal, pacs%flatfield)
    end if

    ! filtering
    select case (filtering_method)

        ! remove mean value in timeline
        case ('none')            
            write (OUTPUT_UNIT,'(a)') 'Removing mean value... '
            start = 1
            do iobs = 1, nobs
                call subtract_meandim1(signal(start:start+obs%slice(iobs)%nsamples-1,:))
                start = start + obs%slice(iobs)%nsamples
            end do


        case ('median')
            write (OUTPUT_UNIT,'(a,i0,a)', advance='no') 'Median filtering (length=', filtering_length, ')... '
            call system_clock(count1, count_rate, count_max)
            start = 1
            do iobs = 1, nobs

                !$omp parallel do
                do idetector = 1, pacs%ndetectors
                    call median_filtering_nocopy(signal(start:start+obs%slice(iobs)%nsamples-1,idetector), filtering_length)
                end do
                !$omp end parallel do
                start = start + obs%slice(iobs)%nsamples
                
            end do
            call system_clock(count2, count_rate, count_max)
            write (OUTPUT_UNIT,'(f7.2,a)') real(count2-count1)/count_rate, 's'

    end select

    ! deglitching
    select case (deglitching_method)

        case ('l2std')
            call deglitch_l2b(pmatrix, nx, ny, signal, mask, deglitching_nsigma, .false., verbose=.true.)

        case ('l2mad')
            call deglitch_l2b(pmatrix, nx, ny, signal, mask, deglitching_nsigma, .true., verbose=.true.)

    end select

    ! back project the timeline
    write(OUTPUT_UNIT,'(a)', advance='no') 'Computing the map... '
    call system_clock(count1, count_rate, count_max)
    call backprojection_weighted(pmatrix, signal, mask, map1d, 0.d0)
    call system_clock(count2, count_rate, count_max)
    write(OUTPUT_UNIT,'(f6.2,a)') real(count2-count1)/count_rate, 's'

    ! write the map as fits file
    write(OUTPUT_UNIT,'(a)') 'Writing FITS file... '
    call ft_write(outfile, reshape(map1d, [nx,ny]), header, status)
    if (status /= 0) go to 999

    call exit(0)

999 stop 'Aborting.'

end program pacs_photproject
